flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/stream/StreamPhysicalJoinRule.scala

```diff
@@ -40,51 +41,39 @@ class StreamPhysicalJoinRule
 
   override def matches(call: RelOptRuleCall): Boolean = {
     val join: FlinkLogicalJoin = call.rel(0)
-    if (!join.getJoinType.projectsRight) {
-      // SEMI/ANTI join always converts to StreamExecJoin now
-      return true
-    }
     val left: FlinkLogicalRel = call.rel(1).asInstanceOf[FlinkLogicalRel]
     val right: FlinkLogicalRel = call.rel(2).asInstanceOf[FlinkLogicalRel]
-    val joinRowType = join.getRowType
-
-    if (left.isInstanceOf[FlinkLogicalSnapshot]) {
-      throw new TableException(
-        "Temporal table join only support apply FOR SYSTEM_TIME AS OF on the right table.")
-    }
 
-    // this rule shouldn't match temporal table join
-    if (right.isInstanceOf[FlinkLogicalSnapshot] ||
-      TemporalJoinUtil.containsTemporalJoinCondition(join.getCondition)) {
+    if (!satisfyRegularJoin(join, right)) {
       return false
     }
 
-    val (windowBounds, remainingPreds) = extractWindowBounds(join)
-    if (windowBounds.isDefined) {
-      return false
-    }
-
-    if (satisfyWindowJoin(join)) {
-      return false
+    // validate the join
+    if (left.isInstanceOf[FlinkLogicalSnapshot]) {
+      throw new TableException(
+        "Temporal table join only support apply FOR SYSTEM_TIME AS OF on the right table.")
     }
 
-    // remaining predicate must not access time attributes
-    val remainingPredsAccessTime = remainingPreds.isDefined &&
-      IntervalJoinUtil.accessesTimeAttribute(remainingPreds.get, joinRowType)
+    // INITIAL_TEMPORAL_JOIN_CONDITION should not appear in physical phase in case which fallback
+    // to regular join
+    checkState(!containsInitialTemporalJoinCondition(join.getCondition))
 
-    val rowTimeAttrInOutput = joinRowType.getFieldList
+    val rowTimeAttrInOutput = join.getRowType.getFieldList
       .exists(f => FlinkTypeFactory.isRowtimeIndicatorType(f.getType))
     if (rowTimeAttrInOutput) {
       throw new TableException(
```

Comment:
The case is different with the above one. 
Before the pr, this exception would be triggered by user because `RelTimeIndicatorConverter` does not cover the case in the old version.
After the pr, it should never happened because `RelTimeIndicatorConverter` already cover the case in the latest version. Here should be `Precondition.checkState` instead of throw `TableException`. The exception should never triggered by users,  otherwise it's a bug. 
I would update the logical here. Thanks.